leader/leader.go

```diff
@@ -124,229 +121,239 @@ func (l *Client) getWatchAtLatestIndex(ctx context.Context, api client.KeysAPI,
 }
 
 // AddWatch starts watching the key for changes and sending them
-// to the valuesC, the watch is stopped
-func (l *Client) AddWatch(key string, retry time.Duration, valuesC chan string) {
-	prefix := fmt.Sprintf("AddWatch(key=%v)", key)
+// to the valuesC.
+// The watch can be stopped either with the specified context or Client.Close
+func (l *Client) AddWatch(ctx context.Context, key string, retry time.Duration, valuesC chan string) {
 	api := client.NewKeysAPI(l.client)
-
-	go func() {
+	entry := log.WithFields(log.Fields{"watch": key})
+	renewCtx, cancel := context.WithCancel(ctx)
+	renew := func() (client.Watcher, *client.Response, error) {
```

Comment:
this code is hard to read - gorotuine and a callback inside with inversion of control? can you simplify this? I don't understand by we need this complex backoff/callback system here.